// wintersmith/lib/cli/index.js
// Generated by CoffeeScript 1.3.3
var globalOptions, logger, main, optimist, path, readJSON, transports, usage, _ref;

path = require('path');

optimist = require('optimist');

_ref = require('./../common'), logger = _ref.logger, transports = _ref.transports, readJSON = _ref.readJSON;

usage = "\nusage: wintersmith [options] [command]\n\ncommands:\n\n  " + 'build'.bold + " [options] - build a site\n  " + 'preview'.bold + " [options] - run local webserver\n  " + 'new'.bold + " <location> - create a new site\n\n  also see [command] --help\n\nglobal options:\n\n  -v, --verbose   show debug information\n  -q, --quiet     only output critical errors\n  -V, --version   output version and exit\n  -h, --help      show help\n";

globalOptions = {
  verbose: {
    alias: 'v'
  },
  quiet: {
    alias: 'q'
  },
  version: {
    alias: 'V'
  },
  help: {
    alias: 'h'
  }
};

main = function() {
  var argv, cmd;
  argv = optimist.options(globalOptions).argv;
  if (argv._[0] != null) {
    try {
      cmd = require("./" + argv._[0]);
    } catch (error) {
      console.log("'" + argv._[0] + "' - no such command");
    }
  }
  if (argv.version) {
    readJSON(path.join(__dirname, '../../package.json'), function(error, result) {
      if (error) {
        return logger.error(error.message, error);
      } else {
        console.log(result.version);
        return process.exit(0);
      }
    });
    return;
  }
  if (argv.help || !cmd) {
    console.log(cmd ? cmd.usage : usage);
    process.exit(0);
  }
  if (argv.verbose) {
    logger.transports.cli.level = 'verbose';
  }
  if (argv.quiet) {
    logger.transports.cli.quiet = true;
  }
  if (cmd) {
    return cmd(optimist.options(globalOptions).options(cmd.options).argv);
  }
};

module.exports.main = main;
